### PR TITLE
sd-boot: LTO support

### PR DESCRIPTION
**Pull request by medhefgo**
I finally figured out how to get LTO support, yay! :joy_cat:

Comparing the output of nm shows it's working (optimizing away functions from other compilation units). And reduces the -O2 binary by about ~6KiB.

So far, it seems to work without any problems (including under secure boot) with gcc.

Currently, lto builts break with gcc+ld.gold on arch linux due to some internal bug in ld.gold, but it did compile earlier on a ubuntu machine, so it should work in theory. Either way, we default to ld.bfd (regardless of distro default).

Building with clang and lto works, but breaks at runtime. It somehow decides not to initialize/optimize away static variables, which breaks console key input and/or secure boot. :cry:
If anyone has an idea as to why clang does this, please let me know.

### CODE REVIEW

**Comment by yuwata** [comment]
build failed on Fedora aarch64:
```
/efi/secure-boot.c.o src/boot/efi/util.c.o src/boot/efi/boot.c.o src/boot/efi/console.c.o src/boot/efi/drivers.c.o src/boot/efi/random-seed.c.o src/boot/efi/shim.c.o src/boot/efi/xbootldr.c.o -lefi -lgnuefi -lgcc
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_PREL32 against symbol `_data' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/lib/gnuefi/aa64/crt0.o: in function `optional_header':
(.text.head+0x5c): dangerous relocation: unsupported relocation
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_PREL32 against symbol `_edata' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/lib/gnuefi/aa64/crt0.o: in function `extra_header_fields':
(.text.head+0x90): dangerous relocation: unsupported relocation
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_PREL32 against symbol `_data' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/lib/gnuefi/aa64/crt0.o: in function `section_table':
(.text.head+0x100): dangerous relocation: unsupported relocation
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_PREL32 against symbol `_data' which may bind externally can not be used when making a shared object; recompile with -fPIC
(.text.head+0x108): dangerous relocation: unsupported relocation
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_PREL32 against symbol `_data' which may bind externally can not be used when making a shared object; recompile with -fPIC
(.text.head+0x12c): dangerous relocation: unsupported relocation
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_PREL32 against symbol `_data' which may bind externally can not be used when making a shared object; recompile with -fPIC
(.text.head+0x134): dangerous relocation: unsupported relocation
/usr/bin/ld.bfd: /usr/lib/gnuefi/aa64/crt0.o: relocation R_AARCH64_ADR_PREL_LO21 against symbol `ImageBase' which may bind externally can not be used when making a shared object; recompile with -fPIC
/usr/lib/gnuefi/aa64/crt0.o: in function `_start':
(.text.head+0x1014): dangerous relocation: unsupported relocation
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Comment by yuwata** [comment]
Please include "Fixes #21258" in the first commit.

**Comment by medhefgo** [comment]
> Please include "Fixes #21258" in the first commit.

Could've sworn that I did. Must've gotten dropped in some rebase at some point. I've taken the liberty to move it intoa separate PR (#21278) since this one might take a while.

To the LTO build: it seems that clang for some reason decides to put static vars into a .bss section we haven't copied into the final efi so far, but only when bulit with -flto.
Now I only need to figure out how to get me a working aarch64 dev environment...

**Comment by poettering** [comment]
meson parts look superficially good to me, but here too I'd leave the merge to our resident meson guru @keszybz.

The third patch definitely lgtm

**Comment by keszybz** [comment]
Let's get #21278 merged first.

**Comment by medhefgo** [comment]
This should work now and does lto with both gcc+bfd/gold and clang+bfd/gold. It even builds now on my ubuntu arm64 vm. The output of nm/objdump tells me that optimization is happening like in x86_64, but mind you I've only compile tested it. But considering that the x86_64 build is running now on my real machine, I see no reason it shouldn't on arm64.

But damn, arm64 emulation is slow af.

**Comment by medhefgo** [comment]
`clang: error: argument unused during compilation: '-T /usr/lib/elf_x86_64_efi.lds' [-Werror,-Wunused-command-line-argument]`
That makes no sense. This is literally passed only to the linking clang command and is needed for it too. Also, I don't get this error on clang-12.

**Comment by yuwata** [comment]
CLANG Build test fails:
```
[27/2323] clang -o src/boot/efi/systemd_boot.so -fuse-ld=bfd -L /usr/lib -nostdlib -T /usr/lib/elf_x86_64_efi.lds -Wl,--build-id=sha1 -Wl,--fatal-warnings -Wl,--no-undefined -Wl,--warn-common -Wl,-Bsymbolic -z nocombreloc /usr/lib/crt0-efi-x86_64.o -pie -Wl,--no-dynamic-linker -Wno-missing-field-initializers -Wno-unused-parameter -Wdate-time -Wendif-labels -Werror=format=2 -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=overflow -Werror=return-type -Werror=shift-count-overflow -Werror=undef -Werror=unused-function -Wfloat-equal -Winit-self -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wwrite-strings '-Wno-error=#warnings' -Wno-string-plus-int -Wno-typedef-redefinition -Wno-gnu-variable-sized-type-not-at-end -fno-stack-protector -fno-strict-aliasing -fpic -Wall -Wextra -Wsign-compare -nostdlib -std=gnu99 -ffreestanding -fshort-wchar -isystem /usr/include/efi -isystem /usr/include/efi/x86_64 -I /home/runner/work/systemd/systemd/src/fundamental -DSD_BOOT -DGNU_EFI_USE_MS_ABI -include src/boot/efi/efi_config.h -include version.h -mno-red-zone -mno-sse -mno-mmx -Werror -ggdb -DEFI_DEBUG src/boot/efi/efivars-fundamental.c.o src/boot/efi/string-util-fundamental.c.o src/boot/efi/sha256.c.o src/boot/efi/assert.c.o src/boot/efi/devicetree.c.o src/boot/efi/disk.c.o src/boot/efi/graphics.c.o src/boot/efi/measure.c.o src/boot/efi/pe.c.o src/boot/efi/secure-boot.c.o src/boot/efi/util.c.o src/boot/efi/boot.c.o src/boot/efi/console.c.o src/boot/efi/drivers.c.o src/boot/efi/random-seed.c.o src/boot/efi/shim.c.o src/boot/efi/xbootldr.c.o -lefi -lgnuefi -lgcc
FAILED: src/boot/efi/systemd_boot.so
clang: error: argument unused during compilation: '-T /usr/lib/elf_x86_64_efi.lds' [-Werror,-Wunused-command-line-argument]
```

**Comment by medhefgo** [comment]
Turns out clang <= 10 is too dumb to pass -T to the linker and then has the audacity to complain about it.

The version compare should technically use efi_cc, but that would mean writing some shell one-liner to detect clang and extract the version, and I don't feel like doing that unless someone insists on this (normaly efi_cc == cc anyway). Also, I would use version_compare but meson insists that `'10.0.0-ubuntu'.version_compare('<=10')` is false.

**Comment by medhefgo** [comment]
So fedora has their gnu-efi heavily patched, making something break at runtime when linked using gcc instead of doing it directly by calling ld. The same happens if I just use the fedora gnu-efi files to build on arch.

The joys of distros doing their own little thing in the name of humanity -.-

File: src/boot/efi/meson.build [hunk 1]
```diff
@@ -258,26 +250,52 @@ foreach arg : get_option('c_args')
...
+        # them seems to work around this.
+        efi_ldflags += [
+                '-Wl,--defsym=_init_array=0',
+                '-Wl,--defsym=_init_array_end=0',
+                '-Wl,--defsym=_fini_array=0',
+                '-Wl,--defsym=_fini_array_end=0',
+                '-Wl,--defsym=__CTOR_LIST__=0',
+                '-Wl,--defsym=__CTOR_END__=0',
+                '-Wl,--defsym=__DTOR_LIST__=0',
+                '-Wl,--defsym=__DTOR_END__=0',
```
Comment (poettering) [inline src/boot/efi/meson.build:298]: constructors are a terrible thing. I don#t want to know what fedora wants with them...


what happens if we define these symbols unconditionally?

File: src/boot/efi/meson.build [hunk 2]
```diff
@@ -306,7 +323,7 @@ foreach tuple : [['systemd_boot.so', systemd_boot_efi_name, systemd_boot_objects
                 tuple[0],
                 input : tuple[2],
                 output : tuple[0],
-                command : [efi_ld, '-o', '@OUTPUT@', efi_ldflags, tuple[2], '-lefi', '-lgnuefi', libgcc_file_name],
+                command : [efi_cc, '-o', '@OUTPUT@', efi_ldflags, tuple[2], '-lefi', '-lgnuefi', '-lgcc'],
```
Comment (poettering) [inline src/boot/efi/meson.build]: what's the -lgcc thing? can you explain? shouldn't this be a separate commit?

**Comment by poettering** [comment]
lgtm, as before, but i think @keszybz as our resident meson export should have a look too

File: src/boot/efi/meson.build (hunk 1 above)
Comment (medhefgo) [reply src/boot/efi/meson.build:298]: That would just mean having some superfluous symbols. Probably. I don't think it even matters what they are since the fedora gnu-efi code only takes their address: https://src.fedoraproject.org/rpms/gnu-efi/blob/rawhide/f/0027-Make-ELF-constructors-and-destructors-work.patch#_602

File: src/boot/efi/meson.build (hunk 2 above)
Comment (medhefgo) [reply src/boot/efi/meson.build]: Basically, the compiler already knows the correct libgcc to link to, so we don't have to do it manually anymore. The `-lgcc` can probably be dropped completely now that I think of it...

Comment (medhefgo) [reply src/boot/efi/meson.build]: Nvm, explicit `-lgcc` is stil needed.

**Comment by keszybz** [comment]
OK, let's merge this. I think there's some risk, but that's what -rc1 releases are for. We'll just ask people to test this.
